### ISSUE TITLE
Terminal/In Game Console Widget

### ISSUE THREAD

**Issue by Pebaz** [emilk/egui#319]
**Is your feature request related to a problem? Please describe.**
I'd like to build an in game console but there does not seem to be a way to do this using the multiline text edit.

**Describe the solution you'd like**
I'd like a widget that has a way to scroll the cursor to the last character in the text box. Additionally, it would be amazing to be able to color specific text blocks in the output for errors and logging, etc.

**Describe alternatives you've considered**
I tried to use a multiline text box but when the buffer is modified (to add the prompt), the cursor stays behind the new text, and there is no way to move the cursor programmatically that I could find.

**Additional context**

```rust
let mut terminal_buffer = String::with_capacity(500);
ui.text_edit_multiline(&mut terminal_buffer);

// Later at the end of the frame:
if <there are two newlines in a row, signaling the user wants to process the command> {
    let command = <scrape out from the last ">>> " to the 2 newlines>;
    // process command
    terminal_buffer += &String::from(">>> ");  // Add new prompt <--- This is where the cursor gets left behind
}
```

Is there a way to move the cursor to the end when the buffer is modified?

**Comment by kotx** [comment]
Maybe try using one multiline TextEdit for scrollback and one singleline TextEdit for input?

**Comment by Pebaz** [comment]
The hard part here is making the multiline widget scroll down to the most recent line added. As currently implemented, there's no way to scroll down to the most recent line.

**Comment by kotx** [comment]
> The hard part here is making the multiline widget scroll down to the most recent line added. As currently implemented, there's no way to scroll down to the most recent line.

Yeah, I'm trying to figure that out too. Currently this is what I have, but resizing the window doesn't work (it just slides back), so the scrollbar is never shown:
```rs
ScrollArea::auto_sized().show(ui, |ui| {
    ui.text_edit_multiline(...);
    ui.scroll_to_cursor(Align::BOTTOM);
});
```

**Comment by Pebaz** [comment]
Oh cool I'll try that and see if that works better than what I'm doing now.

**Comment by kotx** [comment]
It probably doesn't, because it just expands endlessly :P

**Comment by kotx** [comment]
> It probably doesn't, because it just expands endlessly until I figure out how to make it resize properly :P

Seems this is a known issue: https://github.com/emilk/egui/blob/9603bb4f85fbad328bce527ff0850fbc925393d6/egui_demo_lib/src/apps/demo/tests.rs#L424

**Comment by pm100** [comment]
I have built a console window that you describe, its got complicated pretty fast. I was wondering about making it a free standing widget. 

Trying to make it scroll whole lines rather than stoppig with half a line shown

You dont want that scroll to last all the time, becuase it messes with people scrolling back to look at the history
